File: haskell/hssearch/src/HsSearch/SearchOptions.hs
{-# LANGUAGE DeriveGeneric, NoMonomorphismRestriction #-}
module HsSearch.SearchOptions (
    SearchOption(..)
  , SearchOptions(..)
  , getSearchOptions
  , getUsage
  , settingsFromArgs
  , settingsFromFile
  , settingsFromJson) where

import Data.Aeson
import qualified Data.ByteString.Lazy.Char8 as BC
import Data.Char (toLower)
import Data.Either (isLeft, lefts, rights, either)
import Data.List (isPrefixOf, isSuffixOf, sortBy)
import Data.Maybe (fromJust, isJust, listToMaybe)
import GHC.Generics

import HsFind.ArgTokenizer
import HsFind.FileTypes (getFileTypeForName)
import HsFind.FileUtil (getFileString, isFile)
import HsFind.FindOptions (parseDateToUtc)
import HsFind.SortBy (getSortByForName)

import HsSearch.Paths_hssearch (getDataFileName)
import HsSearch.SearchSettings


data SearchOption = SearchOption
  { long :: String
  , short :: Maybe String
  , desc :: String
  } deriving (Show, Eq, Generic)

instance FromJSON SearchOption

newtype JsonSearchOptions
  = JsonSearchOptions {searchoptions :: [SearchOption]}
  deriving (Show, Eq, Generic)

instance FromJSON JsonSearchOptions

data SearchOptions = SearchOptions
  { options :: [SearchOption]
  , argTokenizer :: ArgTokenizer
  } deriving (Show, Eq, Generic)

searchOptionsFile :: FilePath
searchOptionsFile = "searchoptions.json"

getArgTokenizer :: [SearchOption] -> ArgTokenizer
getArgTokenizer jsonOpts =
  ArgTokenizer
    { boolMap = boolMap
    , stringMap = stringMap
    , intMap = intMap
    }
  where
    boolMap = boolShortMap ++ boolLongMap
    stringMap = stringShortMap ++ stringLongMap ++ [("settings-file", "settings-file")]
    intMap = intShortMap ++ intLongMap
    boolLongMap :: [(String, String)]
    boolLongMap = [(long o, long o) | o <- jsonOpts, isBoolOption o]
    boolShortMap :: [(String, String)]
    boolShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isBoolOption o, isJust (short o)]
    stringLongMap :: [(String, String)]
    stringLongMap = [(long o, long o) | o <- jsonOpts, isStringOption o]
    stringShortMap :: [(String, String)]
    stringShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isStringOption o, isJust (short o)]
    intLongMap :: [(String, String)]
    intLongMap = [(long o, long o) | o <- jsonOpts, isIntOption o]
    intShortMap :: [(String, String)]
    intShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isIntOption o, isJust (short o)]
    isBoolOption :: SearchOption -> Bool
    isBoolOption o = long o `elem` map fst boolActions
    isStringOption :: SearchOption -> Bool
    isStringOption o = long o `elem` map fst stringActions
    isIntOption :: SearchOption -> Bool
    isIntOption o = long o `elem` map fst integerActions

getSearchOptions :: IO (Either String SearchOptions)
getSearchOptions = do
  searchOptionsPath <- getDataFileName searchOptionsFile
  searchOptionsJsonString <- getFileString searchOptionsPath
  case searchOptionsJsonString of
    Left e -> return $ Left e
    Right jsonString ->
      case (eitherDecode (BC.pack jsonString) :: Either String JsonSearchOptions) of
        Left e -> return $ Left e
        Right jsonSearchOptions -> return $ Right SearchOptions { options = searchoptions jsonSearchOptions,
                                                                  argTokenizer = getArgTokenizer (searchoptions jsonSearchOptions) }

getUsage :: [SearchOption] -> String
getUsage searchOptions =
  "Usage:\n hssearch [options] -s <searchpattern> <path> [<path> ...]\n\nOptions:\n" ++
  searchOptionsToString searchOptions

getOptStrings :: [SearchOption] -> [String]
getOptStrings = map formatOpts
  where formatOpts SearchOption {long=l, short=Nothing} = getLong l
        formatOpts SearchOption {long=l, short=Just s}  = shortAndLong s l
        getLong l = "--" ++ l
        shortAndLong s l = "-" ++ s ++ "," ++ getLong l

getOptDesc :: SearchOption -> String
getOptDesc SearchOption {desc=""} = error "No description for SearchOption"
getOptDesc SearchOption {desc=d} = d

sortSearchOption :: SearchOption -> SearchOption -> Ordering
sortSearchOption SearchOption {long=l1, short=s1} SearchOption {long=l2, short=s2} =
  compare (shortOrLong s1 l1) (shortOrLong s2 l2)
  where
    shortOrLong Nothing l = l
    shortOrLong (Just s) l = map toLower s ++ "@" ++ l

sortSearchOptions :: [SearchOption] -> [SearchOption]
sortSearchOptions = sortBy sortSearchOption

padString :: String -> Int -> String
padString s len | length s < len = s ++ replicate (len - length s) ' '
                | otherwise      = s

searchOptionsToString :: [SearchOption] -> String
searchOptionsToString searchOptions = 
  unlines $ zipWith formatOptLine optStrings optDescs
  where
    sorted = sortSearchOptions searchOptions
    optStrings = getOptStrings sorted
    optDescs = map getOptDesc sorted
    longest = maximum $ map length optStrings
    formatOptLine o d = " " ++ padString o longest ++ "  " ++ d

data ActionType = BoolActionType
                | StringActionType
                | IntegerActionType
                | UnknownActionType
  deriving (Show, Eq)

type BoolAction = SearchSettings -> Bool -> SearchSettings
type StringAction = SearchSettings -> String -> SearchSettings
type IntegerAction = SearchSettings -> Integer -> SearchSettings

boolActions :: [(String, BoolAction)]
boolActions = [ ("allmatches", \ss b -> ss {firstMatch=not b})
              , ("archivesonly", \ss b -> ss {archivesOnly=b,
                                              searchArchives=b})
              , ("colorize", \ss b -> ss {colorize=b})
              , ("debug", \ss b -> ss {debug=b, verbose=b})
              , ("excludehidden", \ss b -> ss {includeHidden=not b})
              , ("firstmatch", \ss b -> ss {firstMatch=b})
              , ("followsymlinks", \ss b -> ss {followSymlinks=b})
              , ("help", \ss b -> ss {printUsage=b})
              , ("includehidden", \ss b -> ss {includeHidden=b})
              , ("multilinesearch", \ss b -> ss {multiLineSearch=b})
              , ("nocolorize", \ss b -> ss {colorize=not b})
              , ("nofollowsymlinks", \ss b -> ss {followSymlinks=not b})
              , ("noprintdirs", \ss b -> ss {printDirs=not b})
              , ("noprintfiles", \ss b -> ss {printFiles=not b})
              , ("noprintlines", \ss b -> ss {printLines=not b})
              , ("noprintmatches", \ss b -> ss {printResults=not b})
              , ("norecursive", \ss b -> ss {recursive=not b})
              , ("nosearcharchives", \ss b -> ss {searchArchives=not b})
              , ("printdirs", \ss b -> ss {printDirs=b})
              , ("printfiles", \ss b -> ss {printFiles=b})
              , ("printlines", \ss b -> ss {printLines=b})
              , ("printmatches", \ss b -> ss {printResults=b})
              , ("recursive", \ss b -> ss {recursive=b})
              , ("searcharchives", \ss b -> ss {searchArchives=b})
              , ("sort-ascending", \ss b -> ss {sortDescending=not b})
              , ("sort-caseinsensitive", \ss b -> ss {sortCaseInsensitive=b})
              , ("sort-casesensitive", \ss b -> ss {sortCaseInsensitive=not b})
              , ("sort-descending", \ss b -> ss {sortDescending=b})
              , ("uniquelines", \ss b -> ss {uniqueLines=b})
              , ("verbose", \ss b -> ss {verbose=b})
              , ("version", \ss b -> ss {printVersion=b})
              ]

stringActions :: [(String, StringAction)]
stringActions = [ ("encoding", \ss s -> ss {textFileEncoding=s})
                , ("in-archiveext", \ss s -> ss {inArchiveExtensions = inArchiveExtensions ss ++ newExtensions s})
                , ("in-archivefilepattern", \ss s -> ss {inArchiveFilePatterns = inArchiveFilePatterns ss ++ [s]})
                , ("in-dirpattern", \ss s -> ss {inDirPatterns = inDirPatterns ss ++ [s]})
                , ("in-ext", \ss s -> ss {inExtensions = inExtensions ss ++ newExtensions s})
                , ("in-filepattern", \ss s -> ss {inFilePatterns = inFilePatterns ss ++ [s]})
                , ("in-filetype", \ss s -> ss {inFileTypes = inFileTypes ss ++ [getFileTypeForName s]})
                , ("in-linesafterpattern", \ss s -> ss {inLinesAfterPatterns = inLinesAfterPatterns ss ++ [s]})
                , ("in-linesbeforepattern", \ss s -> ss {inLinesBeforePatterns = inLinesBeforePatterns ss ++ [s]})
                , ("linesaftertopattern", \ss s -> ss {linesAfterToPatterns = linesAfterToPatterns ss ++ [s]})
                , ("linesafteruntilpattern", \ss s -> ss {linesAfterUntilPatterns = linesAfterUntilPatterns ss ++ [s]})
                , ("maxlastmod", \ss s -> ss {maxLastMod = parseDateToUtc s})
                , ("minlastmod", \ss s -> ss {minLastMod = parseDateToUtc s})
                , ("out-archiveext", \ss s -> ss {outArchiveExtensions = outArchiveExtensions ss ++ newExtensions s})
                , ("out-archivefilepattern", \ss s -> ss {outArchiveFilePatterns = outArchiveFilePatterns ss ++ [s]})
                , ("out-dirpattern", \ss s -> ss {outDirPatterns = outDirPatterns ss ++ [s]})
                , ("out-ext", \ss s -> ss {outExtensions = outExtensions ss ++ newExtensions s})
                , ("out-filepattern", \ss s -> ss {outFilePatterns = outFilePatterns ss ++ [s]})
                , ("out-filetype", \ss s -> ss {outFileTypes = outFileTypes ss ++ [getFileTypeForName s]})
                , ("out-linesafterpattern", \ss s -> ss {outLinesAfterPatterns = outLinesAfterPatterns ss ++ [s]})
                , ("out-linesbeforepattern", \ss s -> ss {outLinesBeforePatterns = outLinesBeforePatterns ss ++ [s]})
                , ("path", \ss s -> ss {paths = paths ss ++ [s]})
                , ("searchpattern", \ss s -> ss {searchPatterns = searchPatterns ss ++ [s]})
                , ("sort-by", \ss s -> ss {sortResultsBy = getSortByForName s})
                ]

integerActions :: [(String, IntegerAction)]
integerActions = [ ("linesafter", \ss i -> ss {linesAfter = i})
                 , ("linesbefore", \ss i -> ss {linesBefore = i})
                 , ("maxdepth", \ss i -> ss {maxDepth = i})
                 , ("maxlinelength", \ss i -> ss {maxLineLength = i})
                 , ("maxsize", \ss i -> ss {maxSize = i})
                 , ("mindepth", \ss i -> ss {minDepth = i})
                 , ("minsize", \ss i -> ss {minSize = i})
                 ]

updateSettingsFromTokens :: SearchSettings -> SearchOptions -> [ArgToken] -> IO (Either String SearchSettings)
updateSettingsFromTokens settings searchOptions tokens = do
  case tokens of
    [] -> return $ Right settings
    t:ts ->
      case getActionType (name t) of
        BoolActionType ->
          case value t of
            TypeA b -> updateSettingsFromTokens (getBoolAction (name t) settings b) searchOptions ts
            _ -> return $ Left $ "Invalid boolean value for option: " ++ name t
        StringActionType ->
          case value t of
            TypeB s ->
              if name t == "settings-file"
              then do
                settingsEither <- updateSettingsFromFile settings searchOptions s
                case settingsEither of
                  Left e -> return $ Left e
                  Right settings' -> updateSettingsFromTokens settings' searchOptions ts
              else updateSettingsFromTokens (getStringAction (name t) settings s) searchOptions ts
            _ -> return $ Left $ "Invalid string value for option: " ++ name t
        IntegerActionType ->
          case value t of
            TypeC i -> updateSettingsFromTokens (getIntegerAction (name t) settings (toInteger i)) searchOptions ts
            _ -> return $ Left $ "Invalid integer value for option: " ++ name t
        UnknownActionType -> return $ Left $ "Invalid option from updateSettingsFromTokens: " ++ name t
  where
    getActionType :: String -> ActionType
    getActionType a
      | isBoolAction a = BoolActionType
      | isStringAction a = StringActionType
      | a == "settings-file" = StringActionType
      | isIntegerAction a = IntegerActionType
      | otherwise = UnknownActionType
    getBoolAction :: String -> BoolAction
    getBoolAction a = snd $ head $ filter (\(x,_) -> a==x) boolActions
    getStringAction :: String -> StringAction
    getStringAction a = snd $ head $ filter (\(x,_) -> a==x) stringActions
    getIntegerAction :: String -> IntegerAction
    getIntegerAction a = snd $ head $ filter (\(x,_) -> a==x) integerActions
    isBoolAction :: String -> Bool
    isBoolAction a = isJust $ lookup a boolActions
    isStringAction :: String -> Bool
    isStringAction a = isJust $ lookup a stringActions
    isIntegerAction :: String -> Bool
    isIntegerAction a = isJust $ lookup a integerActions

updateSettingsFromJson :: SearchSettings -> SearchOptions -> String -> IO (Either String SearchSettings)
updateSettingsFromJson settings searchOptions jsonStr = do
  let eitherTokens = tokenizeJson (argTokenizer searchOptions) jsonStr
  case eitherTokens of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings searchOptions tokens

settingsFromJson :: SearchOptions -> String -> IO (Either String SearchSettings)
settingsFromJson = updateSettingsFromJson defaultSearchSettings

updateSettingsFromFile :: SearchSettings -> SearchOptions -> FilePath -> IO (Either String SearchSettings)
updateSettingsFromFile settings searchOptions filePath = do
  eitherTokens <- tokenizeFile (argTokenizer searchOptions) filePath
  case eitherTokens of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings searchOptions tokens

settingsFromFile :: SearchOptions -> FilePath -> IO (Either String SearchSettings)
settingsFromFile = updateSettingsFromFile defaultSearchSettings

updateSettingsFromArgs :: SearchSettings -> SearchOptions -> [String] -> IO (Either String SearchSettings)
updateSettingsFromArgs settings searchOptions arguments = do
  case tokenizeArgs (argTokenizer searchOptions) arguments of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings searchOptions tokens

settingsFromArgs :: SearchOptions -> [String] -> IO (Either String SearchSettings)
settingsFromArgs = updateSettingsFromArgs defaultSearchSettings{printResults=True}

File: elixir/exsearch/lib/searchoptions.ex
defmodule ExSearch.SearchOption do
  @moduledoc """
  Documentation for `ExSearch.SearchOption`.
  """

  defstruct short_arg: "", long_arg: "", description: ""

  def new(args), do: __struct__(args)

  def sort_arg(option) do
    if option.short_arg == "" do
      option.long_arg
    else
      "#{String.downcase(option.short_arg)}a#{option.long_arg}"
    end
  end

  def to_arg_string(option) do
    if option.short_arg == "" do
      "--#{option.long_arg}"
    else
      "-#{option.short_arg},--#{option.long_arg}"
    end
  end
end

defmodule ExSearch.SearchOptionsLoader do
  @moduledoc """
  Documentation for `ExSearch.SearchOptionsLoader`.
  """

  def load_options() do
    # Load the search options from the searchoptions.json file.
    search_options_path = ExSearch.Config.search_options_path
    {:ok, json} = File.read(search_options_path)
    search_options = JSON.decode!(json)
    search_options["searchoptions"]
    |> Enum.map(fn o -> ExSearch.SearchOption.new([short_arg: Map.get(o, "short", ""), long_arg: o["long"], description: o["desc"]]) end)
  end
end

defmodule ExSearch.SearchOptions do
  @moduledoc """
  Documentation for `ExSearch.SearchOptions`.
  """

  alias ExFind.ArgTokenizer
  alias ExFind.FileTypes
  alias ExFind.SortBy
  alias ExFind.StringUtil

  alias ExSearch.SearchError
  alias ExSearch.SearchSettings

  require OptionParser

  defstruct options: ExSearch.SearchOptionsLoader.load_options()

  def new(), do: __struct__()

  defp bool_arg_action_map() do
    %{
      allmatches: fn b, settings -> %{settings | first_match: not b} end,
      archivesonly: fn b, settings -> SearchSettings.set_archives_only(settings, b) end,
      colorize: fn b, settings -> %{settings | colorize: b} end,
      debug: fn b, settings -> SearchSettings.set_debug(settings, b) end,
      excludehidden: fn b, settings -> %{settings | include_hidden: not b} end,
      firstmatch: fn b, settings -> %{settings | first_match: b} end,
      followsymlinks: fn b, settings -> %{settings | follow_symlinks: b} end,
      help: fn b, settings -> %{settings | print_usage: b} end,
      includehidden: fn b, settings -> %{settings | include_hidden: b} end,
      multilinesearch: fn b, settings -> %{settings | multi_line_search: b} end,
      nocolorize: fn b, settings -> %{settings | colorize: not b} end,
      nofollowsymlinks: fn b, settings -> %{settings | follow_symlinks: not b} end,
      noprintdirs: fn b, settings -> %{settings | print_dirs: not b} end,
      noprintfiles: fn b, settings -> %{settings | print_files: not b} end,
      noprintlines: fn b, settings -> %{settings | print_lines: not b} end,
      noprintmatches: fn b, settings -> %{settings | print_matches: not b} end,
      noprintresults: fn b, settings -> %{settings | print_results: not b} end,
      noprintusage: fn b, settings -> %{settings | print_usage: not b} end,
      norecursive: fn b, settings -> %{settings | recursive: not b} end,
      nosearcharchives: fn b, settings -> %{settings | search_archives: not b} end,
      printdirs: fn b, settings -> %{settings | print_dirs: b} end,
      printfiles: fn b, settings -> %{settings | print_files: b} end,
      printlines: fn b, settings -> %{settings | print_lines: b} end,
      printmatches: fn b, settings -> %{settings | print_matches: b} end,
      printresults: fn b, settings -> %{settings | print_results: b} end,
      printusage: fn b, settings -> %{settings | print_usage: b} end,
      printversion: fn b, settings -> %{settings | print_version: b} end,
      recursive: fn b, settings -> %{settings | recursive: b} end,
      searcharchives: fn b, settings -> %{settings | search_archives: b} end,
      sort_ascending: fn b, settings -> %{settings | sort_descending: not b} end,
      sort_caseinsensitive: fn b, settings -> %{settings | sort_case_insensitive: b} end,
      sort_casesensitive: fn b, settings -> %{settings | sort_case_insensitive: not b} end,
      sort_descending: fn b, settings -> %{settings | sort_descending: b} end,
      uniquelines: fn b, settings -> %{settings | unique_lines: b} end,
      verbose: fn b, settings -> %{settings | verbose: b} end,
      version: fn b, settings -> %{settings | print_version: b} end
    }
  end

  defp int_arg_action_map() do
    %{
      linesafter: fn i, settings -> %{settings | lines_after: i} end,
      linesbefore: fn i, settings -> %{settings | lines_before: i} end,
      maxdepth: fn i, settings -> %{settings | max_depth: i} end,
      maxlinelength: fn i, settings -> %{settings | max_line_length: i} end,
      maxsize: fn i, settings -> %{settings | max_size: i} end,
      mindepth: fn i, settings -> %{settings | min_depth: i} end,
      minsize: fn i, settings -> %{settings | min_size: i} end
    }
  end

  defp str_arg_action_map() do
    %{
      encoding: fn s, settings -> SearchSettings.set_encoding(settings, s) end,
      in_archiveext: fn s, settings -> SearchSettings.add_extensions(settings, s, :in_archive_extensions) end,
      in_archivefilepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :in_archive_file_patterns) end,
      in_dirpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :in_dir_patterns) end,
      in_ext: fn s, settings -> SearchSettings.add_extensions(settings, s, :in_extensions) end,
      in_filepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :in_file_patterns) end,
      in_filetype: fn s, settings -> %{settings | in_file_types: settings.in_file_types ++ [FileTypes.get_file_type_for_name(s)]} end,
      in_linesafterpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :in_lines_after_patterns) end,
      in_linesbeforepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :in_lines_before_patterns) end,
      linesaftertopattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :lines_after_to_patterns) end,
      linesafteruntilpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :lines_after_until_patterns) end,
      maxlastmod: fn s, settings -> %{settings | max_last_mod: StringUtil.to_datetime(s)} end,
      minlastmod: fn s, settings -> %{settings | min_last_mod: StringUtil.to_datetime(s)} end,
      out_archiveext: fn s, settings -> SearchSettings.add_extensions(settings, s, :out_archive_extensions) end,
      out_archivefilepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :out_archive_file_patterns) end,
      out_dirpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :out_dir_patterns) end,
      out_ext: fn s, settings -> SearchSettings.add_extensions(settings, s, :out_extensions) end,
      out_filepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :out_file_patterns) end,
      out_filetype: fn s, settings -> %{settings | out_file_types: settings.out_file_types ++ [FileTypes.get_file_type_for_name(s)]} end,
      out_linesafterpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :out_lines_after_patterns) end,
      out_linesbeforepattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :out_lines_before_patterns) end,
      path: fn s, settings -> %{settings | paths: settings.paths ++ [s]} end,
      searchpattern: fn s, settings -> SearchSettings.add_pattern(settings, s, :search_patterns) end,
      sort_by: fn s, settings -> %{settings | sort_by: SortBy.get_sort_by_for_name(s)} end
    }
  end

  defp arg_action_maps() do
    {bool_arg_action_map(), int_arg_action_map(), str_arg_action_map()}
  end

  defp get_arg_tokenizer(options, arg_action_maps) do
    {bool_arg_action_map, int_arg_action_map, str_arg_action_map} = arg_action_maps
    %ArgTokenizer{
      options: options,
      bool_opts: Map.keys(bool_arg_action_map),
      int_opts: Map.keys(int_arg_action_map),
      str_opts: Map.keys(str_arg_action_map)
    }
  end

  def update_settings_from_json(settings, json, arg_tokenizer, arg_action_maps) do
    case ArgTokenizer.tokenize_json(json, arg_tokenizer) do
      {:ok, tokens} -> update_settings_from_tokens(settings, tokens, arg_tokenizer, arg_action_maps)
      {:error, message} -> {:error, message}
    end
  end

  def get_settings_from_json(json, options) do
    arg_action_maps = arg_action_maps()
    arg_tokenizer = get_arg_tokenizer(options, arg_action_maps)
    update_settings_from_json(SearchSettings.new(), json, arg_tokenizer, arg_action_maps)
  end

  def get_settings_from_json!(json, options) do
    case get_settings_from_json(json, options) do
      {:error, message} -> raise SearchError, message: message
      {:ok, settings} -> settings
    end
  end

  def update_settings_from_file(settings, json_file, arg_tokenizer, arg_action_maps) do
    case ArgTokenizer.tokenize_file(json_file, arg_tokenizer) do
      {:ok, tokens} -> update_settings_from_tokens(settings, tokens, arg_tokenizer, arg_action_maps)
      {:error, message} -> {:error, message}
    end
  end

  def get_settings_from_file(json_file, options) do
    arg_action_maps = arg_action_maps()
    arg_tokenizer = get_arg_tokenizer(options, arg_action_maps)
    case update_settings_from_file(SearchSettings.new(), json_file, arg_tokenizer, arg_action_maps) do
      {:error, "Unable to parse JSON"} -> {:error, "Unable to parse JSON in settings file: #{json_file}"}
      {:error, message} -> {:error, message}
      {:ok, settings} -> {:ok, settings}
    end
  end

  def get_settings_from_file!(json_file, options) do
    case get_settings_from_file(json_file, options) do
      {:error, message} -> raise SearchError, message: message
      {:ok, settings} -> settings
    end
  end

  def update_settings_from_tokens!(settings, tokens, arg_tokenizer, arg_action_maps) do
    {bool_arg_action_map, int_arg_action_map, str_arg_action_map} = arg_action_maps
    case tokens do
      [] -> settings
      [t | ts] ->
        case t.arg_type do
          :boolean ->
            k = t.name
            v = t.value
            cond do
              Map.has_key?(bool_arg_action_map, k) ->
                update_settings_from_tokens!(Map.get(bool_arg_action_map, k).(v, settings), ts, arg_tokenizer, arg_action_maps)
              true -> raise SearchError, message: "Invalid option: #{k}"
            end
          :integer ->
            k = t.name
            v = t.value
            cond do
              Map.has_key?(int_arg_action_map, k) ->
                update_settings_from_tokens!(Map.get(int_arg_action_map, k).(v, settings), ts, arg_tokenizer, arg_action_maps)
              true -> raise SearchError, message: "Invalid option: #{k}"
            end
          :string ->
            k = t.name
            v = t.value
            cond do
              Map.has_key?(str_arg_action_map, k) ->
                update_settings_from_tokens!(Map.get(str_arg_action_map, k).(v, settings), ts, arg_tokenizer, arg_action_maps)
              k == :settings_file -> case update_settings_from_file(settings, v, arg_tokenizer, arg_action_maps) do
                {:ok, new_settings} -> update_settings_from_tokens!(new_settings, ts, arg_tokenizer, arg_action_maps)
                {:error, message} -> raise SearchError, message: message
              end
              true -> raise SearchError, message: "Invalid option: #{k}"
            end
          :unknown ->
            raise SearchError, message: "Invalid option: #{t.name}"
        end
    end
  end

  def update_settings_from_tokens(settings, tokens, arg_tokenizer, arg_action_maps) do
    try do
      {:ok, update_settings_from_tokens!(settings, tokens, arg_tokenizer, arg_action_maps)}
    rescue
      e in SearchError -> {:error, e.message}
    end
  end

  def update_settings_from_args!(settings, args, options) do
    arg_action_maps = arg_action_maps()
    arg_tokenizer = get_arg_tokenizer(options, arg_action_maps)
    case ArgTokenizer.tokenize_args(args, arg_tokenizer) do
      {:ok, tokens} -> update_settings_from_tokens!(settings, tokens, arg_tokenizer, arg_action_maps)
      {:error, message} -> raise SearchError, message: message
    end
  end

  def update_settings_from_args(settings, args, options) do
    try do
      {:ok, update_settings_from_args!(settings, args, options)}
    rescue
      e in SearchError -> {:error, e.message}
    end
  end

  def get_settings_from_args(args, options) do
    settings = SearchSettings.new([print_results: true])
    update_settings_from_args(settings, args, options)
  end

  def get_settings_from_args!(args, options) do
    case get_settings_from_args(args, options) do
      {:error, message} -> raise SearchError, message: message
      {:ok, settings} -> settings
    end
  end

  defp get_usage_string(options) do
    opt_strings = options
                  |> Enum.sort(fn o1, o2 -> ExSearch.SearchOption.sort_arg(o1) <= ExSearch.SearchOption.sort_arg(o2) end)
                  |> Enum.map(fn o -> {ExSearch.SearchOption.to_arg_string(o), o.description} end)
    longest = Enum.map(opt_strings, fn {opt, _} -> String.length(opt) end) |> Enum.max()
    opt_lines = opt_strings
                |> Enum.map(fn {opt, desc} -> " #{opt}#{String.duplicate(" ", longest - String.length(opt))}  #{desc}" end)
    """
    \nUsage:
     exsearch [options] -s <searchpattern> <path> [<path> ...]

    Options:
    #{Enum.join(opt_lines, "\n")}
    """
  end

  def usage(options) do
    IO.puts(get_usage_string(options))
  end

end
